test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/internal/KtorNetworkTransportTest.kt

```diff
@@ -0,0 +1,137 @@
+
+internal class KtorNetworkTransportTest {
+
+    private lateinit var transport: KtorNetworkTransport
+
+    enum class HTTPMethod(val nativeKey: String) {
+        GET("get"),
+        POST("post"),
+        PATCH("patch"),
+        PUT("put"),
+        DELETE("delete")
+    }
+
+    @BeforeTest
+    fun setUp() {
+        transport = KtorNetworkTransport(
+            timeoutMs = 5000,
+            dispatcher = singleThreadDispatcher("ktor-test")
+        )
+    }
+
+    @Test
+    fun requestSuccessful() {
+        val url = "$BASE_URL/okhttp?success=true"
+
+        // FIXME test fails on native due to method ordering, e.g.:
+        //  1 put (success), 2 put (error)
+        //  1 get (success), 2 post (success), 3 patch (success), 4 put (error)
+        for (method in HTTPMethod.values()) {
+            val body = if (method == HTTPMethod.GET) "" else "{ \"body\" : \"some content\" }"
+
+            val response = transport.sendRequest(
+                method.nativeKey,
+                url,
+                mapOf(),
+                body,
+                true
+            )
+            println("------------> METHOD: $method")
+            assertEquals(200, response.httpResponseCode)
+            assertEquals(0, response.customResponseCode)
+            assertEquals("${method.name}-success", response.body)
+            println("------------> METHOD: $method SUCCESSFUL")
+        }
+    }
+
+    @Test
+    fun requestFailedOnServer() {
+        val url = "$BASE_URL/okhttp?success=false"
+        for (method in HTTPMethod.values()) {
+            val body = if (method == HTTPMethod.GET) "" else "{ \"body\" : \"some content\" }"
+
+            val response = transport.sendRequest(
+                method.nativeKey,
+                url,
+                mapOf(),
+                body,
+                true
+            )
+            assertEquals(500, response.httpResponseCode)
+            assertEquals(0, response.customResponseCode)
+            assertEquals("${method.name}-failure", response.body)
+        }
+    }
+
+    // Make sure that the client doesn't crash if attempting to send invalid JSON
+    // This is mostly a guard against Java crashing if ObjectStore serializes the wrong
+    // way by accident.
+    @Test
+    fun requestSendsIllegalJson() {
+        val url = "$BASE_URL/okhttp?success=true"
+        for (method in HTTPMethod.values()) {
+            val body = if (method == HTTPMethod.GET) "" else "Boom!"
+
+            val response: Response = transport.sendRequest(
+                method.nativeKey,
+                url,
+                mapOf(),
+                body,
+                true
+            )
+            assertEquals(200, response.httpResponseCode)
+            assertEquals(0, response.customResponseCode)
+            assertEquals("${method.name}-success", response.body)
+        }
+    }
+
+//    @Test
+//    fun requestInterrupted() {
```

Comment:
We don't need to test Interrupt pr. se, but in Java this test is here to verify that exceptions from the network layer correctly propagate back to the user.

We should have a similar test for Kotlin. It can probably be done by mocking the correct places. But it is fine to leave a TODO for now.

Reply:
Added an empty `@Ignore`d `errorPropagation` tests with a FIXME and comments about intention.